diff --git a/chromium.spec b/chromium.spec
index b37d37c..816d18c 100644
--- a/chromium.spec
+++ b/chromium.spec
@@ -31,7 +31,7 @@
 	ninja -j %{numjobs} -C '%1' '%2'
 
 # enable|disable headless client build
-%global build_headless 1
+%global build_headless 0
 
 # enable|disable chrome-remote-desktop build
 %global build_remoting 0
@@ -117,7 +117,7 @@
 %global remotingbuilddir out/Remoting
 
 # enable|disable debuginfo
-%global enable_debug 1
+%global enable_debug 0
 # disable debuginfo due to a bug in debugedit on el7
 # error: canonicalization unexpectedly shrank by one character
 # https://bugzilla.redhat.com/show_bug.cgi?id=304121
@@ -128,7 +128,7 @@
 %global debug_package %{nil}
 %global debug_level 0
 %else
-%global debug_level 1
+%global debug_level 0
 # workaround for the error empty file debugsource
 %undefine _debugsource_packages
 %endif
@@ -162,7 +162,7 @@
 %global cfi 0
 %if %{clang}
 %if 0%{?fedora} || 0%{?rhel} > 7
-%global cfi 1
+%global cfi 0
 %endif
 %endif
 
@@ -175,13 +175,13 @@
 
 # enable qt backend for el >= 8 and fedora
 %if 0%{?rhel} >= 8 || 0%{?fedora}
-%global use_qt 1
+%global use_qt 0
 %else
 %global use_qt 0
 %endif
 
 %if 0%{?rhel} > 9 || 0%{?fedora}
-%global use_qt6 1
+%global use_qt6 0
 %else
 %global use_qt6 0
 %endif
@@ -285,7 +285,7 @@
 %endif
 
 Name:	chromium%{chromium_channel}
-Version: 120.0.6099.216
+Version: 122.0.6226.2
 Release: 1%{?dist}
 Summary: A WebKit (Blink) powered web browser that Google doesn't want you to use
 Url: http://www.chromium.org/Home
@@ -446,6 +446,9 @@ patch356: chromium-120-disable-FFmpegAllowLists.patch
 # remove ldflags -Wl,-mllvm,-disable-auto-upgrade-debug-info which is not supported
 Patch357: chromium-120-clang16-disable-auto-upgrade-debug-info.patch
 
+Patch1000: 0001-m122-OT-SVG-hook.patch
+Patch1001: 0002-m122-use-FreeType-for-non-Linux-OT-SVG-support.patch
+
 # upstream patches
 
 # Use chromium-latest.py to generate clean tarball from released build tarballs, found here:
@@ -1029,9 +1032,9 @@ udev.
 
 %if %{clang}
 %if 0%{?rhel} < 8 || 0%{?fedora} < 38
-%patch -P300 -p1 -b .no_matching_constructor
+#%patch -P300 -p1 -b .no_matching_constructor
 %patch -P301 -p1 -b .workaround_clang-SkColor4f
-%patch -P302 -p1 -b .workaround_clang_bug-structured_binding
+#%patch -P302 -p1 -b .workaround_clang_bug-structured_binding
 %patch -P303 -p1 -b .typename
 %patch -P304 -p1 -b .string-convert
 %patch -P306 -p1 -b .assert
@@ -1044,17 +1047,17 @@ udev.
 %endif
 %endif
 
-%patch -P310 -p1 -b .missing-header-files
+#%patch -P310 -p1 -b .missing-header-files
 %patch -P311 -p1 -b .clang-warnings
 %patch -P312 -p1 -b .fstack-protector-strong
 
-%patch -P351 -p1 -b .mnemonic-error
+#%patch -P351 -p1 -b .mnemonic-error
 
 %if %{disable_bti}
 %patch -P352 -p1 -b .workaround_for_crash_on_BTI_capable_system
 %endif
 
-%patch -P353 -p1 -b .gn-workaround-atspi
+#%patch -P353 -p1 -b .gn-workaround-atspi
 %patch -P354 -p1 -b .revert-split-threshold-for-reg-with-hint
 %if ! %{use_custom_libcxx}
 %patch -P355 -p1 -b .nullptr_t-without-namespace-std
@@ -1119,6 +1122,17 @@ sed -i 's/getenv("CHROME_VERSION_EXTRA")/"Fedora Project"/' chrome/common/channe
 # Fix hardcoded path in remoting code
 sed -i 's|/opt/google/chrome-remote-desktop|%{crd_path}|g' remoting/host/setup/daemon_controller_delegate_linux.cc
 
+# third_party/rust-src is 4.3G and we had enable_rust=false
+rm -rf third_party/rust-src
+# 1.9GB
+rm -rf third_party/llvm
+
+tar -czpf ${GITHUB_WORKSPACE}/before.tgz skia third_party/ots third_party/skia third_party/blink/renderer/platform
+ls -l ${GITHUB_WORKSPACE}/before.tgz
+
+%patch -P1000 -p1 -b .ot-svg1
+%patch -P1001 -p1 -b .ot-svg2
+
 %build
 # utf8 issue on epel7, Internal parsing error 'ascii' codec can't
 # decode byte 0xe2 in position 474: ordinal not in range(128)
@@ -1126,6 +1140,8 @@ sed -i 's|/opt/google/chrome-remote-desktop|%{crd_path}|g' remoting/host/setup/d
 export LANG=en_US.UTF-8
 %endif
 
+rm %{SOURCE0}
+
 # reduce warnings
 %if %{clang}
 FLAGS=' -Wno-deprecated-declarations -Wno-unknown-warning-option -Wno-unused-command-line-argument'
@@ -1222,7 +1238,7 @@ CHROMIUM_CORE_GN_DEFINES+=' use_lld=false'
 %endif
 
 # disable rust, it's only using for testing
-CHROMIUM_CORE_GN_DEFINES+=' enable_rust=false'
+CHROMIUM_CORE_GN_DEFINES+=' enable_rust=false enable_cxx=false enable_rust_qr=false'
 
 CHROMIUM_CORE_GN_DEFINES+=' use_sysroot=false disable_fieldtrial_testing_config=true'
 
@@ -1379,6 +1395,10 @@ tools/gn/bootstrap/bootstrap.py --gn-gen-args="$CHROMIUM_CORE_GN_DEFINES $CHROMI
 mkdir -p %{builddir} && cp -a %{_bindir}/gn %{builddir}/
 %endif
 
+perl -pi -e 's(^.+components/qr_code_generator:unit_tests.+$)()' components/BUILD.gn
+perl -pi -e 's(^.+components/qr_code_generator:qr_code_generator_features.+$)()' chrome/services/qrcode_generator/public/cpp/BUILD.gn
+perl -pi -e 's(^.+components/qr_code_generator.+$)()' chrome/services/qrcode_generator/BUILD.gn
+
 %{builddir}/gn --script-executable=%{chromium_pybin} gen --args="$CHROMIUM_CORE_GN_DEFINES $CHROMIUM_BROWSER_GN_DEFINES" %{builddir}
 
 %if %{build_headless}
@@ -1394,6 +1414,10 @@ mkdir -p %{builddir} && cp -a %{_bindir}/gn %{builddir}/
 %build_target %{headlessbuilddir} headless_shell
 %endif
 
+%build_target %{builddir} third_party/blink/renderer/platform
+
+exit $?
+
 %build_target %{builddir} chrome
 %build_target %{builddir} chrome_sandbox
 %build_target %{builddir} chromedriver
